lib/rubocop/cop/rspec/overwriting_setup.rb

```diff
@@ -0,0 +1,55 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module RSpec
+      # Checks if there is a let/subject that overwrites an existing one.
+      #
+      # @example
+      #   # bad
+      #     let(:foo) { bar }
+      #     let(:foo) { baz }
+      #
+      #     subject(:foo) { bar }
+      #     let(:foo) { baz }
+      #
+      #     let(:foo) { bar }
+      #     let!(:foo) { baz }
+      #
+      #   # good
+      #     subject(:test) { something }
+      #     let(:foo) { bar }
+      #     let(:baz) { baz }
+      #     let!(:other) { other }
+      class OverwritingSetup < Cop
+        MSG = '`%{name}` is already defined.'.freeze
+
+        def_node_matcher :setup?, <<-PATTERN
+          (block (send nil {:let :let! :subject} (sym $_)) ...)
+        PATTERN
+
+        def on_block(node)
+          return unless example_group?(node)
+
+          _describe, _args, body = *node
+
+          find_duplicates(body) do |duplicate, name|
+            add_offense(duplicate, :expression, format(MSG, name: name))
+          end
+        end
+
+        private
+
+        def find_duplicates(node)
+          setup_expressions = []
```

Comment:
Not that I imagine it would influence the performance much either way, but `Set.new` here might be more appropriate.

(idk if it actually leads to better code / if you want to use it, but there's also an `#add?` method on `Set` which returns nil if the element was already in the set so you could do `yield child, name unless setup_expressions.add?(name)`.)